### PR TITLE
introduce cancel to `UpgradeStep`

### PR DESCRIPTION
**Pull request by rgildein**
Introduce cancel to `UpgradeStep`, which will cancel all sub-steps. The `cancel` can by run with `safe=True` (default value), which will keep all pending task running and prevent new one to be triggered. The `safe=False` will cancel also all pending `asyncio.Task` (running), which can cause stopping in middle of step and break cloud.

With these changes I also used coroutine as input of `UpgradeStep` instead of function and parameters. This helps with catching calling some coroutine with wrong arguments, e.k. linting cat catch it. However this change required to add warning ignore, since coroutines can be un-awaited in case of only planning.

FYI: We could not define `UpgradeStep` as `asyncio.Task` , since `asyncio.Task` is scheduled to be run as soon as possible if it's run in loop already. We can use it only if whole planning will be not asynchronous. 

related: #103 (preparation for proper handling of `KeyboardInterrupt`

### CODE REVIEW

File: tests/unit/steps/test_steps.py [hunk 1]
```diff
@@ -13,104 +13,246 @@
...
+    """Test UpgradeStep adding sub steps."""
+    exp_sub_steps = 3
+    plan = UpgradeStep(description="plan")
+    for i in range(exp_sub_steps):
+        plan.add_step(UpgradeStep(description=f"sub-step-{i}"))
+
+    assert len(plan.sub_steps) == exp_sub_steps
+
+
+def test_step_cancel():
```
Comment (gabrielcocenza) [inline tests/unit/steps/test_steps.py:134]:
```suggestion
def test_step_cancel_safe():
```

File: tests/unit/steps/test_steps.py [hunk 2]
```diff
@@ -13,104 +13,246 @@
...
+    sub_steps[0].sub_steps = [UpgradeStep(description=f"sub-0.{i}") for i in range(3)]
+
+    plan.cancel()
+
+    assert plan.canceled is True
+    assert all(step.canceled is True for step in sub_steps)
+    assert all(step.canceled for step in sub_steps[0].sub_steps)
+
+
+def test_step_unsafe_cancel():
```
Comment (gabrielcocenza) [inline tests/unit/steps/test_steps.py:148]:
```suggestion
def test_step_cancel_unsafe():
```

File: tests/unit/steps/test_steps.py [hunk 3]
```diff
@@ -13,104 +13,246 @@
...
+        ),
     ],
 )
-def test_upgrade_step__eq__(description, function, parallel, expected_result):
-    upgrade_step = UpgradeStep(description="test", function=None, parallel=False)
-    other_upgrade_step = UpgradeStep(description=description, function=function, parallel=parallel)
-    assert (upgrade_step == other_upgrade_step) is expected_result
+async def test_step_full_run(sub_steps, exp_order, parallel):
+    """Test to simulate running full plan with steps."""
+    # warnings.filterwarnings("ignore", message="coroutine '.*' was never awaited")
```
Comment (gabrielcocenza) [inline tests/unit/steps/test_steps.py:232]: is this comment missing to remove?

File: tests/unit/steps/test_steps.py [hunk 4]
```diff
@@ -13,104 +13,246 @@
...
+    assert hash(("test hash", False, coro)) == hash(step)
+
+
+@pytest.mark.parametrize(
+    "description, parallel, args",
+    [("test", False, ()), ("test description", True, ("name", 1, 2))],
+)
+def test_step_eq(description, parallel, args):
+    """Test UpgradeStep comparison."""
+    # coro = mock_coro(*args)
```
[the remark this answers is not in the document]
Comment (rgildein) [reply tests/unit/steps/test_steps.py:73]: Thanks, I was debugging the warning errors.

File: tests/unit/steps/test_steps.py [hunk 5]
```diff
@@ -13,104 +13,246 @@
...
+    step_1 = UpgradeStep(description, parallel, mock_coro(*args))
+    step_2 = UpgradeStep(description, parallel, mock_coro(*args))
+    # define step with different coro
+    step_3 = UpgradeStep(description, parallel, mock_coro(unique_arg=True))
+
+    assert step_1 == step_2
+    assert step_1 != step_3
+    assert step_1 != 1
```
[the remark this answers is not in the document]
Comment (rgildein) [reply tests/unit/steps/test_steps.py:81]: Trying only if `return NotImplemented` is working fine if `not isinstance(other, UpgradeStep)`.

File: tests/unit/steps/test_steps.py (hunk 3 above)
Comment (rgildein) [reply tests/unit/steps/test_steps.py:232]: Yes again, thanks.

**Comment by rgildein** [comment]
> I personally had a bad time trying to understand what is the `_task` introduced in the UpgradeStep and I wish it had more explanation for future maintainability.

Do you have any suggestion?
The main reason is that instead awaiting the `core` we create task which can be canceled later + we are defining name of for task so we can have watcher later checking if task is still running or something like that.

File: cou/steps/__init__.py [hunk 6]
```diff
@@ -112,3 +125,64 @@ def __repr__(self) -> str:
...
+
+        self._canceled = True
+        logger.debug("canceled: %s", self)
+
+    async def run(self) -> Any:
+        """Run the UpgradeStep coroutine.
+
+        :return: Result of the coroutine.
+        :rtype: Any
+        :raises CanceledUpgradeStep: If step has already been cancelled.
```
Comment (agileshaw) [inline cou/steps/__init__.py:174]:
```suggestion
        :raises CanceledUpgradeStep: If step has already been canceled.
```

**Review by agileshaw — CHANGES_REQUESTED** [review]
Just two small suggestions inline. Overall I like the new approach and am looking forward to seeing everything comes together.

**Review by gabrielcocenza — APPROVED** [review]

**Review by agileshaw — APPROVED** [review]
LGTM

**Review by agileshaw — APPROVED** [review]

**Review by gabrielcocenza — APPROVED** [review]

**Pull request merged**